app/models/chargeback/consumption_without_rollups.rb

```diff
@@ -33,20 +33,29 @@ def parents_determining_rate
        MiqEnterprise.my_enterprise].compact
     end
 
-    def none?(_metric)
-      true # No, values except for fixed (RateDetail.fixed?)
-    end
-
-    def max(_metric)
-      raise NotImplementedError # Unreachable code since none?==true
-    end
-
-    def avg(_metric)
-      raise NotImplementedError # Unreachable code since none?==true
+    def none?(metric)
+      current_value(metric).nil?
     end
 
     def chargeback_fields_present
       1 # Yes, charge this interval as fixed_compute_*_*
     end
+
+    def current_value(metric)
+      # Return the last seen allocation for charging purposes.
+      @value ||= {}
+      @value[metric] ||= case metric
+                         when 'derived_vm_numvcpus' # Allocated CPU count
+                           resource.hardware.try(:cpu_total_cores)
+                         when 'derived_memory_available'
+                           resource.hardware.try(:memory_mb)
+                         when 'derived_vm_allocated_disk_storage'
+                           resource.allocated_disk_storage
+                         end
+      @value[metric]
+    end
+    alias avg current_value
```

Comment:
You are right, downloading the data is costly. And I am working on bringing the database churn to absolute minimum. This, however, needs to be understood in broader context.

I view world as a wrestle between local optimization and global optimization. I strive to optimize the Chargeback subsystem as a whole. That however makes the changes harder to understand.

I am not focusing on single query (although I insist this one is already in good shape) -- but I focus on entry points. Most of the queries we did in darga or euwe were not needed.

To avoid queries we don't need, we have to tell what queries affect the reports results and what queries are dropped later on. To be able to tell what affects what, we need to structure code that represents problem domain as closely as possible.

Consumption object is represents chunk of time of given resource (i.e. week of Metric rollups)

For each consumption and metric tuple, we need to calculate `max` or `avg` only
 - if the given metric is represented on the report (as column)
 - OR if parent (group) metric is represented on the report
 - OR if the cost (or parent group) of the given metric is represented on the report and the applied rate is not zero and rate is not fixed

That means the focus of optimization should be first to *not to run `avg` and `max` unless needed.

I have opened #12885 for that a month ago.

-----------------------
Side note: we cannot use `avg` method from SQL, our denominator may differ.

We could re-use max method, one day. Not now. If we optimize this too much locally today our ability to optimize globally would be hindered.

-----------------------

Now let's discuss the `max` and the `avg` methods. These created on ConsumptionWithRollups and ConsumptionWithoutRollups classes. With rollups, these methods have the real meaning, as described above. This pr is however about the case without rollups.

ConsumptionWithoutRollups today applies only to Hyper-V resources. Hyper-V do not have Metrics collection at all. As per pm comments, in this case users are interested in charging per allocation only. The assumption was made consciously, that current_value is good approximation of real-time allocation.